### PR TITLE
types: Convert provider to typescript

### PR DESCRIPTION
**Pull request by PlayMa256**
**What kind of change does this PR introduce?**
Types: Adding types to Provider component

**Did you add tests for your changes?**
N/A

**Does this PR introduce a breaking change?**
No

### CODE REVIEW

File: packages/reakit/src/Provider/Provider.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import * as React from "react";
+import * as PropTypes from "prop-types";
+import {
+  Provider as ConstateProvider,
+  ProviderProps as ConstateProps
+} from "constate";
+import { ThemeProvider, ThemeProviderProps } from "../styled";
+
+interface ComponentProps
```
Comment (diegohaz) [inline packages/reakit/src/Provider/Provider.tsx]: Let's call it `ProviderProps` and export it the same way we're doing on https://github.com/reakit/reakit/pull/220#issuecomment-419767826.

File: packages/reakit/src/Provider/Provider.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  extends ConstateProps<{ [key: string]: any }>,
```
Comment (diegohaz) [inline packages/reakit/src/Provider/Provider.tsx]: Maybe it's better to receive a generic `S` in `ProviderProps`, like `interface ProviderProps<S extends { [key: string]: any }>` and pass `S` to `ConstateProps`.

```ts
export interface ProviderProps<S extends { [key: string]: any }> extends ConstateProps<S> ...
```

File: packages/reakit/src/Provider/Provider.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    ThemeProviderProps<Object> {
+  children: React.ReactNode;
+}
+
+const Provider: React.SFC<ComponentProps> = ({ theme, children, ...props }) => (
```
Comment (diegohaz) [inline packages/reakit/src/Provider/Provider.tsx]: I don't know how to put the generic `S` here. Maybe transforming it into a class?

File: packages/reakit/src/Provider/Provider.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  <ConstateProvider {...props}>
+    {theme ? <ThemeProvider theme={theme}>{children}</ThemeProvider> : children}
+  </ConstateProvider>
+);
+
+// @ts-ignore
+Provider.propTypes = {
```
Comment (diegohaz) [inline packages/reakit/src/Provider/Provider.tsx]: Does it throw any errors?

File: packages/reakit/src/Provider/Provider.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+} from "constate";
+import { ThemeProvider, ThemeProviderProps } from "../styled";
+
+export interface ProviderProps<S extends { [key: string]: any }>
+  extends ConstateProps<S>,
+    ThemeProviderProps<Object> {
+  children: React.ReactNode;
+}
+
+class Provider<S> extends React.Component<ProviderProps<S>> {
```
Comment (matheus1lva) [inline packages/reakit/src/Provider/Provider.tsx:16]: I hope i got it right, otherwise let me know!

**Pull request merged**